src/Facebook/FileUpload/FacebookResumableUploader.php

```diff
@@ -0,0 +1,201 @@
+     */
+    protected $maxTransferTries;
+
+    /**
+     * @param FacebookApp $app
+     * @param FacebookClient $client
+     * @param string $accessToken
+     * @param int $maxTransferTries
+     */
+    public function __construct(FacebookApp $app, FacebookClient $client, $accessToken, $maxTransferTries = 5)
+    {
+        $this->app = $app;
+        $this->client = $client;
+        $this->accessToken = $accessToken;
+        $this->maxTransferTries = $maxTransferTries;
+    }
+
+    /**
+     * Upload - phase start
+     *
+     * @param $endpoint
+     * @param $filePath
+     * @return FacebookTransferChunk
+     *
+     * @throws FacebookSDKException
+     */
+    public function start($endpoint, $filePath)
+    {
+        $fileSize = filesize($filePath);
+
+        $startReqParams = [
+            'upload_phase' => 'start',
+            'file_size' => $fileSize,
+        ];
+
+        $request = new FacebookRequest(
+            $this->app,
+            $this->accessToken,
+            'POST',
+            $endpoint,
+            $startReqParams
+        );
+
+        $res= $this->client->sendRequest($request)->getDecodedBody();
+
+        $firstTransferChunk = new FacebookTransferChunk(
+            $filePath,
+            $res['upload_session_id'],
+            $res['video_id'],
+            $res['start_offset'],
+            $res['end_offset'] - $res['start_offset']
+        );
+
+        return $firstTransferChunk;
+    }
+
+    /**
+     * Upload - phase transfer
+     *
+     * @param $endpoint
+     * @param FacebookTransferChunk $chunk
+     * @return FacebookTransferChunk
+     *
+     * @throws FacebookResumableUploadException
+     * @throws FacebookSDKException
+     */
+    public function transfer($endpoint, FacebookTransferChunk $chunk)
+    {
+        $transReqParams = [
+            'upload_phase' => 'transfer',
+            'upload_session_id' => $chunk->getUploadSessionId(),
+            'start_offset' => $chunk->getStartOffset(),
+            'video_file_chunk' => $chunk,
+        ];
+
+        $maxTransferTries = $this->maxTransferTries;
+
+        $request = new FacebookRequest(
+            $this->app,
+            $this->accessToken,
+            'POST',
+            $endpoint,
+            $transReqParams
+        );
+
+        while ($maxTransferTries > 0) {
+            try {
+                $res= $this->client->sendRequest($request)->getDecodedBody();
+
+                $nextTransferChunk = new FacebookTransferChunk(
+                    $chunk->getFilePath(),
+                    $chunk->getUploadSessionId(),
+                    $chunk->getVideoId(),
+                    $res['start_offset'],
+                    $res['end_offset'] - $res['start_offset']
+                );
+
+                return $nextTransferChunk;
+            } catch (FacebookSDKException $e) {
+                $preException = $e->getPrevious();
+                if ($preException instanceof FacebookResumableUploadException) {
```

Comment:
Not sure I get it. @yguedidi , are you suggesting to refactor the code like this?
```
                  $preException = $e->getPrevious();
                 if (! $preException instanceof FacebookResumableUploadException) {
                     throw $e;
                 }
                 if (--$maxTransferTries <= 0) {
                     $resumeContext = new FacebookResumeContext(
                          $endpoint,
                          $this->accessToken,
                          $chunk
                      );
                      $preException->setResumeContext($resumeContext);
                       throw $e;
                  }
```